// iracing-api/lib/types/stats.js
import * as z from 'zod';
import { HelmetSchema } from './common';
import { BasicLiverySchema } from './results';
export const CarsDrivenSchema = z.object({
    carId: z.number(),
    carName: z.string(),
});
export const StatsTrackSchema = z.object({
    trackId: z.number(),
    trackName: z.string(),
    configName: z.string(),
});
export const BestSchema = z.object({
    track: StatsTrackSchema,
    eventType: z.string(),
    bestLapTime: z.number(),
    subsessionId: z.number(),
    endTime: z.string(),
    seasonYear: z.number(),
    seasonQuarter: z.number(),
});
export const MemberBestsSchema = z.object({
    carsDriven: z.array(CarsDrivenSchema),
    bests: z.array(BestSchema),
    custId: z.number(),
    carId: z.number(),
});
export const MemberDivisionSchema = z.object({
    division: z.number(),
    projected: z.boolean(),
    eventType: z.number(),
    success: z.boolean(),
    seasonId: z.number(),
});
export const RecentRaceTrackSchema = z.object({
    trackId: z.number(),
    trackName: z.string(),
});
export const RecentRaceSchema = z.object({
    seasonId: z.number(),
    seriesId: z.number(),
    seriesName: z.string(),
    carId: z.number(),
    carClassId: z.number(),
    livery: BasicLiverySchema,
    licenseLevel: z.number(),
    sessionStartTime: z.string(),
    winnerGroupId: z.number(),
    winnerName: z.string(),
    winnerHelmet: HelmetSchema,
    winnerLicenseLevel: z.number(),
    startPosition: z.number(),
    finishPosition: z.number(),
    qualifyingTime: z.number(),
    laps: z.number(),
    lapsLed: z.number(),
    incidents: z.number(),
    clubPoints: z.number(),
    points: z.number(),
    strengthOfField: z.number(),
    subsessionId: z.number(),
    oldSubLevel: z.number(),
    newSubLevel: z.number(),
    oldiRating: z.number(),
    newiRating: z.number(),
    track: RecentRaceTrackSchema,
    dropRace: z.boolean(),
    seasonYear: z.number(),
    seasonQuarter: z.number(),
    raceWeekNum: z.number(),
});
export const MemberRecentRacesSchema = z.object({
    races: z.array(RecentRaceSchema),
    custId: z.number(),
});
export const WorldRecordsChunkInfoSchema = z.object({
    chunk_size: z.number(),
    num_chunks: z.number(),
    rows: z.number(),
    base_download_url: z.string(),
    chunk_file_names: z.array(z.string()),
});
export const WorldRecordsDataSchema = z.object({
    success: z.boolean(),
    car_id: z.number(),
    track_id: z.number(),
    chunk_info: WorldRecordsChunkInfoSchema,
    last_updated: z.string(),
});
export const WorldRecordsSchema = z.object({
    type: z.literal('stats_world_records'),
    data: WorldRecordsDataSchema,
});
export const StatSchema = z.object({
    categoryId: z.number(),
    category: z.string(),
    starts: z.number(),
    wins: z.number(),
    top5: z.number(),
    poles: z.number(),
    avgStartPosition: z.number(),
    avgFinishPosition: z.number(),
    laps: z.number(),
    lapsLed: z.number(),
    avgIncidents: z.number(),
    avgPoints: z.number(),
    winPercentage: z.number(),
    top5Percentage: z.number(),
    lapsLedPercentage: z.number(),
    totalClubPoints: z.number(),
});
export const MemberCareerSchema = z.object({
    stats: z.array(StatSchema),
    custId: z.number(),
});
export const FavoriteTrackSchema = z.object({
    trackId: z.number(),
    trackName: z.string(),
    configName: z.string(),
    trackLogo: z.string(),
});
export const FavoriteCarSchema = z.object({
    carId: z.number(),
    carName: z.string(),
    carImage: z.string(),
});
export const StatsSchema = z.object({
    starts: z.number(),
    wins: z.number(),
    top5: z.number(),
    avgStartPosition: z.number(),
    avgFinishPosition: z.number(),
    laps: z.number(),
    lapsLed: z.number(),
    favoriteCar: FavoriteCarSchema,
    favoriteTrack: FavoriteTrackSchema,
});
export const MemberRecapSchema = z.object({
    year: z.number(),
    stats: StatsSchema,
    success: z.boolean(),
    season: z.null(),
    custId: z.number(),
});
export const ThisYearSchema = z.object({
    numOfficialSessions: z.number(),
    numLeagueSessions: z.number(),
    numOfficialWins: z.number(),
    numLeagueWins: z.number(),
});
export const MemberSummarySchema = z.object({
    thisYear: ThisYearSchema,
    custId: z.number(),
});
export const YearlyStatSchema = z.intersection(StatSchema, z.object({
    year: z.number(),
}));
export const MemberYearlyStatsSchema = z.object({
    stats: z.array(YearlyStatSchema),
    custId: z.number(),
});
// Params
export const GetMemberBestsParamsSchema = z.object({
    customerId: z.number().optional(),
    carId: z.number().optional(),
});
export const GetMemberCareerParamsSchema = z.object({
    customerId: z.number().optional(),
});
export const GetMemberDivisionParamsSchema = z.object({
    seasonId: z.number(),
    eventType: z.number(),
});
export const GetMemberRecentRacesParamsSchema = z.object({
    customerId: z.number().optional(),
});
export const GetMemberRecapParamsSchema = z.object({
    customerId: z.number().optional(),
    year: z.number().optional(),
    season: z.number().optional(),
});
export const GetMemberSummaryParamsSchema = z.object({
    customerId: z.number().optional(),
});
export const GetMemberYearlyStatsParamsSchema = z.object({
    customerId: z.number().optional(),
});
export const GetDriverSeasonStandingsParamsSchema = z.object({
    seasonId: z.number(),
    carClassId: z.number(),
    clubId: z.number().optional(),
    division: z.number().optional(),
    raceWeekNumber: z.number().optional(),
});
export const getSupersessionSeasonStandingsParamsSchema = z.object({
    seasonId: z.number(),
    carClassId: z.number(),
    clubId: z.number().optional(),
    division: z.number().optional(),
    raceWeekNumber: z.number().optional(),
});
export const GetTeamSeasonStandingsParamsSchema = z.object({
    seasonId: z.number(),
    carClassId: z.number(),
    raceWeekNumber: z.number().optional(),
});
export const GetTimeTrialSeasonStandingsParamsSchema = z.object({
    seasonId: z.number(),
    carClassId: z.number(),
    clubId: z.number().optional(),
    division: z.number().optional(),
    raceWeekNumber: z.number().optional(),
});
export const GetTimeTrialSeasonResultsParamsSchema = z.object({
    seasonId: z.number(),
    carClassId: z.number(),
    raceWeekNumber: z.number(),
    clubId: z.number().optional(),
    division: z.number().optional(),
});
export const GetQualifySeasonResultsParamsSchema = z.object({
    seasonId: z.number(),
    carClassId: z.number(),
    clubId: z.number().optional(),
    division: z.number().optional(),
    raceWeekNumber: z.number().optional(),
});
export const GetWorldRecordsParamsSchema = z.object({
    carId: z.number(),
    trackId: z.number(),
    seasonYear: z.number().optional(),
    seasonQuarter: z.number().optional(),
});
